rust/src/lib.rs

```diff
@@ -602,78 +603,550 @@ pub extern "C" fn zcashlc_get_address(
+///   encoded Unified Address.
+/// - Call [`zcashlc_string_free`] to free the memory associated with the returned pointer
+///   when done using it.
+#[no_mangle]
+pub extern "C" fn zcashlc_get_sapling_receiver_for_unified_address(
+    ua: *const c_char,
+) -> *mut c_char {
+    let res = catch_panic(|| {
+        let ua_str = unsafe { CStr::from_ptr(ua).to_str()? };
+
+        let (network, ua) = match ZcashAddress::try_from_encoded(ua_str) {
+            Ok(addr) => addr
+                .convert::<(_, UnifiedAddressParser)>()
+                .map_err(|e| format_err!("Not a Unified Address: {}", e)),
+            Err(e) => return Err(format_err!("Invalid Zcash address: {}", e)),
+        }?;
+
+        if let Some(addr) = ua.0.sapling() {
+            Ok(
+                CString::new(ZcashAddress::from_sapling(network, addr.to_bytes()).encode())?
+                    .into_raw(),
+            )
+        } else {
+            Err(format_err!(
+                "Unified Address doesn't contain a Sapling receiver"
+            ))
+        }
+    });
+    unwrap_exc_or_null(res)
+}
+
+/// Returns true when the provided address decodes to a valid Sapling payment address for the
+/// specified network, false in any other case.
+///
+/// # Safety
+///
+/// - `address` must be non-null and must point to a null-terminated UTF-8 string.
+/// - The memory referenced by `address` must not be mutated for the duration of the function call.
+#[no_mangle]
+pub extern "C" fn zcashlc_is_valid_shielded_address(
     address: *const c_char,
     network_id: u32,
 ) -> bool {
     let res = catch_panic(|| {
         let network = parse_network(network_id)?;
-        let addr = CStr::from_ptr(address).to_str()?;
-        Ok(is_valid_shielded_address(&addr, &network))
+        let addr = unsafe { CStr::from_ptr(address).to_str()? };
+        Ok(is_valid_shielded_address(addr, &network))
     });
     unwrap_exc_or(res, false)
 }
 
 fn is_valid_shielded_address(address: &str, network: &Network) -> bool {
-    match RecipientAddress::decode(network, &address) {
+    match RecipientAddress::decode(network, address) {
         Some(addr) => match addr {
             RecipientAddress::Shielded(_) => true,
-            RecipientAddress::Transparent(_) => false,
+            RecipientAddress::Transparent(_) | RecipientAddress::Unified(_) => false,
         },
         None => false,
     }
 }
 
-/// Returns true when the address is valid and transparent.
-/// Returns false in any other case
+enum AddressType {
+    Sprout,
+    P2pkh,
+    P2sh,
+    Sapling,
+    Unified,
+}
+
+struct AddressMetadata {
+    network: zcash_address::Network,
+    addr_type: AddressType,
+}
+
+#[derive(Debug)]
+enum Void {}
+
+impl TryFromAddress for AddressMetadata {
+    /// This instance produces no errors; ideally this would be the void
+    /// (uninhabitable) type but I'm not sure how to do that in Rust.
```

Comment:
```suggestion
    /// This instance produces no errors.
```
The empty `enum` is the idiomatic way of doing this in Rust, I think.

Reply:
Looks like there’s also https://docs.rs/void/latest/void/, which has some useful related functions, like extracting the guaranteed value out of a `Result` that has `Void` on one side and `unreachable` (which is generally called `absurd` … but I like `unreachable`).